Add support for rocJPEG tracing

sputhala-amd opened the pull request:
Add rocDecode API Tracing support using domain rocjpeg_api in ROCPROFSYS_ROCM_DOMAINS.
Modify existing videodecode and jpegdecode ctest to verify API tracing
Print Perfetto values for easy debugging in verbose mode
Fix cmake error coming from videodecode building to convert it to a warning if rocDecode is not found.

sputhala-amd committed:
Add support for rocJPEG tracing

* rocDecode API Tracing support using domain rocjpeg_API
* Modify existing videodecode and jpegdecode ctest to verify API tracing
* print perfetto values for easy debuging in verbose mode

dgaliffiAMD commented on source/lib/core/categories.hpp:
```diff
@@ -107,7 +107,8 @@ ROCPROFSYS_DEFINE_CATEGORY(category, rocm_smi_power, ROCPROFSYS_CATEGORY_ROCM_SM
 ROCPROFSYS_DEFINE_CATEGORY(category, rocm_smi_vcn_activity, ROCPROFSYS_CATEGORY_ROCM_SMI_VCN_ACTIVITY, "device_vcn_activity", "VCN Activity of a GPU device")
 ROCPROFSYS_DEFINE_CATEGORY(category, rocm_smi_jpeg_activity, ROCPROFSYS_CATEGORY_ROCM_SMI_JPEG_ACTIVITY, "device_jpeg_activity", "JPEG Activity of a GPU device")
-ROCPROFSYS_DEFINE_CATEGORY(category, rocm_rocdecode_api, ROCPROFSYS_CATEGORY_ROCM_ROCDECODE_API, "rocm_rocdecode_api", "ROCm ROCDecode API")
+ROCPROFSYS_DEFINE_CATEGORY(category, rocm_rocdecode_api, ROCPROFSYS_CATEGORY_ROCM_ROCDECODE_API, "rocm_rocdecode_api", "ROCm RocDecode API")
+ROCPROFSYS_DEFINE_CATEGORY(category, rocm_rocjpeg_api, ROCPROFSYS_CATEGORY_ROCM_ROCJPEG_API, "rocm_rocjpeg_api", "ROCm RocJPEG API")
```
Can you move the rocdecode_api and rocjpeg_api categories up to L102 (just below marker_api) to keep all of the rocprofiler domains grouped together?

dgaliffiAMD commented on source/lib/core/categories.hpp:
```diff
@@ -174,6 +175,7 @@ using name = perfetto_category<Tp...>;
         ROCPROFSYS_PERFETTO_CATEGORY(category::rocm_smi_vcn_activity),                   \
         ROCPROFSYS_PERFETTO_CATEGORY(category::rocm_smi_jpeg_activity),                  \
         ROCPROFSYS_PERFETTO_CATEGORY(category::rocm_rocdecode_api),                      \
+        ROCPROFSYS_PERFETTO_CATEGORY(category::rocm_rocjpeg_api),                        \
```
Can you move the rocdecode_api and rocjpeg_api categories up to L169 (just below marker_api) to keep all of the rocprofiler domains grouped together?

dgaliffiAMD commented on source/lib/core/config.cpp:
```diff
@@ -1885,7 +1885,9 @@ get_use_vaapi_tracing()
-           domain_list.end();
+               domain_list.end() ||
+           std::find(domain_list.begin(), domain_list.end(), "rocjpeg_api") !=
+               domain_list.end();  // Check if decode_api or encode_api is present
```
```suggestion
               domain_list.end();  // Check if decode_api or jpeg_api are present
```

dgaliffiAMD commented on source/lib/core/rocprofiler-sdk.cpp:
```diff
@@ -406,11 +407,13 @@ get_callback_domains()
+#    if(ROCPROFILER_VERSION_MAJOR == 0 && ROCPROFILER_VERSION_MINOR >= 7) ||             \
         ROCPROFILER_VERSION_MAJOR >= 1
-        else if(itr == "rocdecode_api")
+        else if(itr == "rocdecode_api" || itr == "rocjpeg_api")
```
If either rocdecode or rocjpeg are added to the config, then enable both? Are we sure that we want to do it this way?
If we want to create a grouping, then let's create a unique name. "decode_api" maybe?

sputhala-amd replied:
That is a mistake ... I should not be enabling both.

However, I did think about the grouping. Currently we are keeping the domain names same as returned by the SDK where the rocdecode_api and rocjpeg_api are defined separately. For within rocprofiler-systems we can create them as a group. That also means we explicitly mention this in our documentation. Either ways are fine, we keep this same as rocprofiler-sdk or have a grouping within our code.

I am keeping the domain names separate as returned by SDK for now and making the edits to cover the error. But please let me know if grouping is better and I can modify the code accordingly.

dgaliffiAMD replied:
The "else" case should take care of this situation, then. You shouldn't need to write your own "else-if" block, if you're using the same string as the ones provided by the SDK. 
Later, we can talk to the domain experts to ask if a grouping would be useful and convenient.

sputhala-amd replied:
Thanks. Yes I understood how the "else-if" is enough.

dgaliffiAMD commented on source/lib/core/rocprofiler-sdk.cpp:
```diff
@@ -481,6 +484,15 @@ get_buffered_domains()
         }
+#    if(ROCPROFILER_VERSION_MAJOR == 0 && ROCPROFILER_VERSION_MINOR >= 7) ||             \
+        ROCPROFILER_VERSION_MAJOR >= 1
+        else if(itr == "rocdecode_api" || itr == "rocjpeg_api")
```
If either rocdecode or rocjpeg are added to the config, then enable both? Are we sure that we want to do it this way?
If we want to create a grouping, then let's create a unique name. "decode_api" maybe?

dgaliffiAMD commented on source/lib/rocprof-sys-user/rocprofiler-systems/categories.h:
```diff
@@ -58,6 +58,7 @@ extern "C"
         ROCPROFSYS_CATEGORY_ROCM_SMI_POWER,
         ROCPROFSYS_CATEGORY_ROCM_SMI_MEMORY_USAGE,
         ROCPROFSYS_CATEGORY_ROCM_ROCDECODE_API,
+        ROCPROFSYS_CATEGORY_ROCM_ROCJPEG_API,
```
Can you move the rocdecode_api and rocjpeg_api categories up to L54 (just below marker_api) to keep all of the rocprofiler domains grouped together?

sputhala-amd committed:
Fix cmake errors from missing rocDecode

Convert the cmake errors to warning if we are building the tests with ROCPROFSYS_BUILD_TESTING=ON
if rocDecode pacakges are missing.
